onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -0,0 +1,721 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <deque>
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/nchwc_transformer.h"
+#include "core/mlas/inc/mlas.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+class NchwcTransformerImpl {
+ public:
+  NchwcTransformerImpl(Graph& graph) noexcept : graph_(graph) {}
+
+  void Transform(Node& node);
+  void Finalize(bool& modified);
+
+  static constexpr int kNchwcDims = 4;
+
+ private:
+  // Associate the following state with each created NCHWc output keyed off the
+  // original NodeArg.
+  struct NchwcArgument {
+    struct Shape {
```

Comment:
'Shape' is somewhat strongly associated with the dims of a tensor, whereas here it's the node outputs. Could there maybe be some documentation added to this struct to explain what it represents a little more clearly as I had to read through the code in order to make sense of it.

Reply:
Added more comments.